optimum/neuron/utils/generation_utils.py

```diff
@@ -0,0 +1,188 @@
+from .import_utils import is_neuronx_distributed_available
+
+import torch
+import torch_xla.core.xla_model as xm
+from transformers import GenerationConfig
+
+from .misc import args_and_kwargs_to_kwargs_only
+
+if is_neuronx_distributed_available():
+    from neuronx_distributed.parallel_layers import parallel_state
+
+
+def move_dict_args_to_device(kwargs: Dict[str, Any], device: str = "cpu") -> Dict[str, Any]:
+    """
+    Takes keyword arguments which will be passed to a model's forward function
+    and moves its values to `device` if
+    they are of type `torch.Tensor`. If the key is a dictionary it does the same to the
+    respective values.
+    Args:
+        kwargs: (`Dict[str, Any]`):
+            The kwargs to be passed to the models forward function.
+        device: (`str`, defaults to `cpu`):
+            The target device to which tensors should be moved.
+    Returns:
+        `Dict[str, Any]`: The kwargs dict with its tensors moved to `device`.
+    """
+
+    def needs_move(src_device, tgt_device):
+        return src_device != tgt_device
+
+    for k, v in kwargs.items():
+        # Handle nested dicts
+        if isinstance(v, dict):
+            for k_, v_ in v.items():
+                if isinstance(v_, torch.Tensor):
+                    if needs_move(v_.device, device):
+                        v[k_] = v_.to(device=device)
+
+        # Handle tensor types
+        elif isinstance(v, torch.Tensor):
+            if needs_move(v.device, device):
+                kwargs[k] = v.to(device=device)
+
+        # Handle past_key_value tuples
+        elif k == "past_key_values":
+            if v is not None:
+                new_past_key_values = ()
+                for layer_past in v:
+                    new_layer_past = ()
+                    for past_state in layer_past:
+                        if needs_move(past_state.device, device):
+                            new_layer_past += (past_state.to(device=device),)
+                        else:
+                            new_layer_past += (past_state,)
+                    new_past_key_values += (new_layer_past,)
+                kwargs[k] = new_past_key_values
+
+    return kwargs
+
+
+def pad_input_ids_for_general_sampling(
+    input_ids: torch.Tensor,
+    num_padding_values: int,
+    pad_token_id: int
+) -> torch.Tensor:
+    """
+    Pads `input_ids` with `num_padding_values` padding tokens along the second dimension.
+    Args:
+        input_ids (`torch.Tensor`):
+            Input ids to be padded.
+        num_padding_values (`int`):
+            Number of padding values to add.
+        pad_token_id (`int`):
+            Token ID of padding token.
+    Returns:
+        `torch.Tensor`: Padded `input_ids`.
+    """
+    bsz = input_ids.size(0)
+    input_ids = torch.cat(
+        [input_ids, torch.ones((bsz, num_padding_values), device=input_ids.device, dtype=torch.long) * pad_token_id], 1
+    )
+    return input_ids
+
+
+def get_fwd_for_general_sampling(
```

Comment:
Since eventually you are overloading forward, you could as well have done it statically in your mixin, which for now contains a single overloaded method (generate).

Reply:
According to HuggingFace's implementation, `GenerationMixin` class is the base class of `PretraineModel` classes. I cannot call `super().forward` in `GeneralNeuronGenerationMixin` class because its base class doesn't implement any `forward` method.  I can access the model's forward using `self.forward` in GeneralNeuronGenerationMixin` class because the child model classes never implement `generate` function.

Reply:
You're right: `forward` is provided by the child class whose `GenerationMixin` inheritance is replaced by this new class, so your only option is to overload it dynamically.
I am wondering what happens when you call generate a second time: since you never restore the original forward, will it not result in wrapping forward again and again ?
I would personally have tried doing it in the constructor: the model class always needs to call super() in its constructor because of the `torch.nn.Module` inheritance.

Reply:
I'm working on more testings right now, and I found similar issue regarding wrapping the forward. Once we replaced the forward method, we need to restore it at the end of the generation in general. Otherwise, the following training steps will brake. There are probably two ways to fix it:
1. Make the new forward general enough for both training and evaluation, and wrap it once in the constructor.
2. In the generation function, wrap the forward and then restore it after generation is done.

I'll explore both and share updates.

Reply:
If you want to wrap temporarily the forward, instead of patching, you could also use the pytorch built-in hooks.
I personally like wrapping them in a context like illustrated here:
https://github.com/huggingface/quanto/blob/main/quanto/quantization/calibrate.py

Reply:
It seems like we the `kwargs` which we need to update will not be passed to the hook, so I think maybe we should not use the hooks in this case. See https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/module.py#L189-L190

Reply:
Ah yes, you're right: these hooks are rather designed for modules whose forward method is usually simpler than for models.